### PR TITLE
max custom redeem algo

### PR DESCRIPTION
**Pull request by tbrent**
This touched a lot of files because of breaking apart `FacadeRead` into two facades, but the meaningful part of the diff is just the new `FacadeMint` contract and the new test: `Should calculate customRedeem portions correctly`. 

Open to better names for the `FacadeMint`. Issue seemed like a worse word to use as a suffix, and we've used suffixes elsewhere.

### CODE REVIEW

**Comment by tbrent** [comment]
After approval I'll deploy

File: contracts/facade/FacadeMint.sol
Comment (pmckelvy1) [inline contracts/facade/FacadeMint.sol:210]: comment is wrong. seems we're walking forward instead

File: test/Facade.test.ts
Comment (pmckelvy1) [inline test/Facade.test.ts:377]: seems like we should also test at least 1 or 2 scenarios that are a little more complex than this?

**Review by pmckelvy1 — CHANGES_REQUESTED** [review]

**Comment by tbrent** [comment]
@akshatmittal @pmckelvy1 this algo turns out to be insufficient. Going either solely forwards or backwards does not work all the time. I think the fastest simple algo is `O(n^2)`: you have to pass over each basket nonce and choose the one with the max number of baskets given current available balances, and you have to do this for every single choice of next basket nonce to use. 

I could imagine discovering there's a `O(n*log(n)) algo somewhere out there if you do a sort on some clever computed metric, but a simple `O(n)` algo does not exist pretty sure.

**Comment by tbrent** [comment]
One not so crazy option is to limit the search to the past `n` basket nonces, where `n` is something like 3 or 5. Or at least, we'd take this as an argument and leave it to the off-chain component to decide which `n` to use.

**Comment by tbrent** [comment]
This is absolutely a sign that this problem is best-solved fully from off-chain, though.

**Comment by akshatmittal** [comment]
I think so too, this should be off-chain ideally. Seems my doubts were reasonable, thanks for looking into it!

**Comment by tbrent** [comment]
I went to start writing the `O(n^2)` solution and it very quickly got past the stack limit. We should close out this PR and not use it. This also means we don't have to split up `FacadeRead` just yet.

**Pull request closed without merging**
